### DISCUSSION TITLE
Assigning a value to a uctype array

### DISCUSSION THREAD

**Discussion by 3phreal** [micropython/micropython#9553] [General]
Hi 

I am trying to create a uctype structure in memory containing an array. But it seems the only way I can write values to this array is by indexing each element separately. 

```
import uctypes
TEMPLATE = {
    "Value"  : (0 | uctypes.ARRAY, 6 | uctypes.UINT8),
    "Value2" : 6 | uctypes.UINT8
}
# Make a new buffer
buffer = bytearray(6)
# Get the address 
buffer_ptr = uctypes.addressof(buffer)

my_struct = uctypes.struct(buffer_ptr, TEMPLATE, uctypes.BIG_ENDIAN)
```

I had hoped I could assign it with a byte array like this. 

`my_struct.Value = bytearray([1,2,3,4,5,6])`

However, this gives me an error. It seems I'm only able to assign values by indexing the source also

```

for y, x in enumerate([0x00, 0x50, 0xB6, 0x0D, 0x35, 0x9F]):
    my_struct.MAC_DST[y] = x

```

Is this correct understood or have I made an error? I would prefer I could assign it directly
I have no problem assigning a value directly to "Value2"

Regards

**Comment by 3phreal** [comment]
I found that i could use the following syntax and it enabled me to make a direct assignment

```
my_struct.Value[:] = bytearray([1,2,3,4,5,6])
```

**Reply by jimmo** [reply]
Yes, this is the right way to do it.

`my_struct.Value = bytearray([1,2,3,4,5,6])` would be replacing the `.Value` attribute on `my_struct`, whereas using the slice assignment you're updating its contents.

To avoid an allocation, it's more efficient to write:

`my_struct.Value[:] = b'\x01\x02\x03\x04\x05\x06'`

because the string will be a const object, whereas the bytearray-of-list will need to allocate a list and then a bytearray every time the code is run. With the byte string, it's allocated once at compile time.
